charon/src/name_matcher/mod.rs

```diff
@@ -0,0 +1,242 @@
+use std::cmp::Ordering;
+
+use itertools::{EitherOrBoth, Itertools};
+use serde::{Deserialize, Serialize};
+
+use crate::ast::*;
+
+mod parser;
+
+pub use Pattern as NamePattern;
+
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Pattern {
+    elems: Vec<PatElem>,
+}
+
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+enum PatElem {
+    /// An identifier, optionally with generic arguments. E.g. `std` or `Box<_>`.
+    Ident {
+        name: String,
+        generics: Vec<PatTy>,
+        /// For pretty-printing only: whether this is the name of a trait.
+        is_trait: bool,
```

Comment:
Is this only used for pretty printing, or also for matching?
For the Charon-ML side: we should have the same, but should also use it for matching (if `true`, we can only match a trait impl).

Reply:
I'm not sure how useful that is: there cannot be a trait with the same name as a type. The fact that we ignore this means we can support both the `impl Trait<..> for Self` syntax and the `Trait<Self, ..>` syntaxes.

Reply:
Well, I guess with wildcards there can be :thinking:

Reply:
Actually you're probably right: it may be useless. But I was mostly thinking about consistency/sanity...